src/libs/FileSorter.cpp

```diff
@@ -0,0 +1,204 @@
+/*
+ * FileSorter.cpp
+ *
+ *  Created on: Sep 29, 2016
+ *      Author: mmoore
+ */
+
+#include "FileSorter.h"
+#include "DirHandle.h"
+#include "FileStream.h"
+
+#include <new>
+#include <malloc.h>
+#include <string.h>
+
+
+FileSorter::FileSorter(string dir_path, __file_sort_filter_fn_t callback)
+{
+    this->directory_path = dir_path;
+    this->error = false;
+    this->file_info_array = NULL;
+    this->file_count = 0;
+    this->total_file_count = 0;
+    this->file_index = -1;
+    this->filter_callback = callback;
+
+    open_directory(dir_path);
+}
+
+FileSorter::~FileSorter()
+{
+    delete_file_info_array();
+}
+
+int FileSorter::compare_dir(file_info_t* file_a, file_info_t* file_b)
+{
+    return strcmp(file_a->file_name, file_b->file_name);
+}
+
+size_t FileSorter::count_files_in_folder(string dir_path)
+{
+    struct dirent *p;
+    uint16_t count = 0;
+    DIR* folder = opendir(dir_path.c_str());
+    if ( folder != NULL ) {
+        while ( (p = readdir(folder)) != NULL ) {
+            if ( (this->filter_callback == NULL) || this->filter_callback(p) ) {
+                count++;
+            }
+        }
+        closedir(folder);
+        return count;
+    }
+    return 0;
+}
+
+void FileSorter::sort_directory(void)
+{
+    // sort the array.
+    qsort(this->file_info_array, this->file_count, sizeof(file_info_t), (__compar_fn_t)compare_dir);
+    this->file_index = -1;
+}
+
+void FileSorter::delete_file_info_array(void)
+{
+    if ( this->file_info_array != NULL) {
+        // file names were allocated previously with strdup - free them now.
+        for ( size_t i=0; i<this->file_count; i++ ) {
+            free(this->file_info_array[i].file_name);
+        }
+
+        // free the array.
+        delete[] this->file_info_array;
+    }
+}
+
+void FileSorter::open_directory(string dir_path)
+{
+    // reset all member state variables.
+    this->directory_path = dir_path;
+    this->error = false;
+    this->file_count = 0;
+    this->total_file_count = 0;
+    this->file_index = -1;
+    delete_file_info_array();
+
```

Comment:
delete_file_info_array() is called after setting file_count to 0.
This is a memory leak, I think.